### PR TITLE
Enrich validate nested mappings

### PR DESCRIPTION
**Pull request by jbaiera**
This PR ensures that fields retained in an enrich index from a source are not contained under a nested field. It additionally makes sure that key fields exist, and that value fields are checked if they are present. The policy runner test has also been expanded with some faulty mapping test cases.

### CODE REVIEW

File: x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java [hunk 1]
```diff
@@ -268,32 +581,68 @@ public void onFailure(Exception e) {
...
+                                                  Long createTime) {
+        ClusterService clusterService = getInstanceFromNode(ClusterService.class);
+        IndexNameExpressionResolver resolver = getInstanceFromNode(IndexNameExpressionResolver.class);
+        return new EnrichPolicyRunner(policyName, policy, listener, clusterService, client(), resolver, () -> createTime,
+            randomIntBetween(1, 10000));
+    }
+
+    private ActionListener<PolicyExecutionResult> createTestListener(final CountDownLatch latch,
+                                                                     final Consumer<Exception> exceptionConsumer) {
+        return new ActionListener<>() {
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java]: maybe use `LatchedActionListener` and `ActionListener#wrap(...)` instead?

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java [hunk 2]
```diff
@@ -109,19 +114,74 @@ private void validateMappings(final GetIndexResponse getIndexResponse) {
...
+        assert fieldName != null && !fieldName.isEmpty() : "Field name cannot be null or empty";
+        String[] fieldParts = fieldName.split("\\.");
+        StringBuilder parent = new StringBuilder();
+        Map<?, ?> currentField = properties;
+        boolean onRoot = true;
+        for (String fieldPart : fieldParts) {
+            // Ensure that the current field is of object type only (not a nested type or a non compound field)
+            Object type = currentField.get("type");
+            if (type != null) {
+                throw new ElasticsearchException(
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java]: an object field can be specifically configured with: `type: object`, so we need to check for this here

File: x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java [hunk 3]
```diff
@@ -212,46 +181,390 @@ public void testRunnerMultiSource() throws Exception {
...
+                        .startObject("properties")
+                            .startObject("field1")
+                                .field("type", "keyword")
+                            .endObject()
+                            .startObject("field2")
+                                .field("type", "integer")
+                            .endObject()
+                            .startObject("field3")
+                                .field("type", "keyword")
+                            .endObject()
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java]: maybe also add a test that has a mapping with two nested object fields (level1.level2.*)?

File: x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java [hunk 4]
```diff
@@ -303,6 +652,6 @@ public void onFailure(Exception e) {
         ShardSegments shard = shardSegments.getShards()[0];
         assertThat(shard.getSegments().size(), is(equalTo(1)));
         Segment segment = shard.getSegments().iterator().next();
-        assertThat(segment.getNumDocs(), is(equalTo(3)));
+        assertThat(segment.getNumDocs(), is(equalTo(expectedDocs)));
```
Comment (martijnvg) [inline x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java]: maybe also add a thing test that configures object fields with a type (`type: object`)?

**Review by martijnvg — COMMENTED** [review]
This looks good. I left a few comments.

**Review by martijnvg — APPROVED** [review]
LGTM

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java [hunk 5]
```diff
@@ -109,19 +114,74 @@ private void validateMappings(final GetIndexResponse getIndexResponse) {
...
+            } catch (ElasticsearchException e) {
+                throw new ElasticsearchException(
+                        "Enrich policy execution for [{}] failed while validating field mappings for index [{}]",
+                        e, policyName, sourceIndex);
+            }
+        }
+    }
+
+    private void validateField(Map<?, ?> properties, String fieldName, boolean fieldRequired) {
+        assert fieldName != null && !fieldName.isEmpty() : "Field name cannot be null or empty";
```
Comment (hub-cap) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java]: `assert Strings.isEmpty(fieldName) == false: "..";`

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java [hunk 6]
```diff
@@ -109,19 +114,74 @@ private void validateMappings(final GetIndexResponse getIndexResponse) {
...
+        String[] fieldParts = fieldName.split("\\.");
+        StringBuilder parent = new StringBuilder();
+        Map<?, ?> currentField = properties;
+        boolean onRoot = true;
+        for (String fieldPart : fieldParts) {
+            // Ensure that the current field is of object type only (not a nested type or a non compound field)
+            Object type = currentField.get("type");
+            if (type != null && !"object".equals(type)) {
```
Comment (hub-cap) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java]: `"object".equals(type) == false`

File: x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java [hunk 7]
```diff
@@ -268,32 +471,422 @@ public void onFailure(Exception e) {
 
         // Validate Mapping
         Map<String, Object> mapping = enrichIndex.getMappings().get(createdEnrichIndex).get("_doc").sourceAsMap();
+        logger.info(mapping);
```
Comment (hub-cap) [inline x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java]: is this meant to be here?

**Review by hub-cap — APPROVED** [review]
minor nits

**Pull request merged**